File: backend/app/models/document.py
import uuid
from datetime import datetime, timezone
from sqlalchemy import DateTime, ForeignKey, Integer, String
from sqlalchemy.orm import Mapped, mapped_column, relationship
from app.models.base import Base


class Document(Base):
    __tablename__ = "documents"

    id: Mapped[str] = mapped_column(
        String, primary_key=True, default=lambda: str(uuid.uuid4())
    )
    workspace_id: Mapped[str] = mapped_column(
        String, ForeignKey("workspaces.id", ondelete="CASCADE"), nullable=False, index=True
    )
    uploaded_by_id: Mapped[str] = mapped_column(
        String, ForeignKey("users.id"), nullable=False, index=True
    )
    filename: Mapped[str] = mapped_column(String(255), nullable=False)
    file_size: Mapped[int] = mapped_column(Integer, nullable=False)
    mime_type: Mapped[str] = mapped_column(String(100), default="application/pdf", nullable=False)
    storage_path: Mapped[str] = mapped_column(String(500), nullable=False)
    status: Mapped[str] = mapped_column(
        String(50), default="UPLOADED", nullable=False
    )  # UPLOADED, PARSING, EMBEDDING, INDEXING, READY, FAILED

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=lambda: datetime.now(timezone.utc), nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
        nullable=False,
    )

    workspace = relationship("Workspace")
    uploaded_by = relationship("User")
    doc_metadata = relationship(
        "DocumentMetadata",
        back_populates="document",
        uselist=False,
        cascade="all, delete-orphan",
        lazy="selectin",
    )


class DocumentMetadata(Base):
    __tablename__ = "document_metadata"

    id: Mapped[str] = mapped_column(
        String, primary_key=True, default=lambda: str(uuid.uuid4())
    )
    document_id: Mapped[str] = mapped_column(
        String, ForeignKey("documents.id", ondelete="CASCADE"), nullable=False, unique=True
    )
    page_count: Mapped[int | None] = mapped_column(Integer, nullable=True)
    title: Mapped[str | None] = mapped_column(String(500), nullable=True)
    author: Mapped[str | None] = mapped_column(String(255), nullable=True)
    checksum: Mapped[str | None] = mapped_column(String(64), nullable=True)

    document = relationship("Document", back_populates="doc_metadata")

File: backend_rust/src/models/document.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use sqlx::FromRow;
use uuid::Uuid;

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct Document {
    pub id: Uuid,
    pub workspace_id: Uuid,
    pub uploaded_by_id: Uuid,
    pub filename: String,
    pub file_size: i64,
    pub mime_type: String,
    pub storage_path: String,
    pub status: String,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
    pub extracted_text: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct DocumentMetadata {
    pub id: Uuid,
    pub document_id: Uuid,
    pub page_count: Option<i32>,
    pub title: Option<String>,
    pub author: Option<String>,
    pub checksum: Option<String>,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DocumentResponse {
    pub id: Uuid,
    pub workspace_id: Uuid,
    pub uploaded_by_id: Uuid,
    pub filename: String,
    pub file_size: i64,
    pub mime_type: String,
    pub storage_path: String,
    pub status: String,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
    pub doc_metadata: Option<DocumentMetadataResponse>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DocumentMetadataResponse {
    pub id: Uuid,
    pub document_id: Uuid,
    pub page_count: Option<i32>,
    pub title: Option<String>,
    pub author: Option<String>,
    pub checksum: Option<String>,
}

impl From<Document> for DocumentResponse {
    fn from(doc: Document) -> Self {
        Self {
            id: doc.id,
            workspace_id: doc.workspace_id,
            uploaded_by_id: doc.uploaded_by_id,
            filename: doc.filename,
            file_size: doc.file_size,
            mime_type: doc.mime_type,
            storage_path: doc.storage_path,
            status: doc.status,
            created_at: doc.created_at,
            updated_at: doc.updated_at,
            doc_metadata: None,
        }
    }
}

impl From<DocumentMetadata> for DocumentMetadataResponse {
    fn from(meta: DocumentMetadata) -> Self {
        Self {
            id: meta.id,
            document_id: meta.document_id,
            page_count: meta.page_count,
            title: meta.title,
            author: meta.author,
            checksum: meta.checksum,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DocumentStatusUpdate {
    pub status: String,
}